packages/keystone/runtimeStats.js

```diff
@@ -0,0 +1,171 @@
+const ENABLED_VAR_NAME = 'RUNTIME_STATS_ENABLE'
+
+const IS_BUILD_PHASE = conf.PHASE === 'build'
+const IS_WORKER_PROCESS = conf.PHASE === 'worker'
+
+const logger = getLogger('runtime-stats')
+
+class RuntimeStatsMiddleware {
+    constructor ({
+        statsUrl = '/api/runtime-stats',
+    } = {}) {
+        this.statsUrl = statsUrl
+    }
+
+    async prepareMiddleware ({ keystone }) {
+        if (conf[ENABLED_VAR_NAME] !== 'true') {
+            logger.info({ msg: 'runtime stats disabled' })
+            return
+        }
+
+        if (!conf[X_TARGET_OPTIONS_VAR_NAME]) {
+            logger.warn({
+                msg: 'There are no options for x-target header. All queries will relate to the `other` group.',
+                data: { howToFix: `Add the ${X_TARGET_OPTIONS_VAR_NAME} variable to env. For example: 'condo-app,billing,cc-app'` },
+            })
+        }
+
+        const runtimeStats = {
+            activeRequestsIds: new Set(),
+            activeRequestsCountByType: {},
+            activeRequestsCountByTarget: {},
+            totalRequestsCount: 0,
+            totalRequestsCountByType: {},
+            totalRequestsCountByTarget: {},
+        }
+
+        if (!IS_BUILD_PHASE && !IS_WORKER_PROCESS) {
+            setInterval(function sendRuntimeMetrics () {
+                const { activeRequestsIds, ...otherStats } = runtimeStats
+
+                logger.info({
+                    msg: 'current values',
+                    runtimeStats: { activeRequestsIds: Array.from(activeRequestsIds.keys()), ...otherStats },
+                })
+
+                metrics.gauge({ name: 'runtimeStats.requestsCount.total', value: runtimeStats.totalRequestsCount })
+
+                for (const type of Object.keys(runtimeStats.totalRequestsCountByType)) {
+                    metrics.gauge({
+                        name: `runtimeStats.requestsCount.type.${camelCase(type)}.total`,
+                        value: runtimeStats.totalRequestsCountByType[type],
+                    })
+                }
+
+                for (const target of Object.keys(runtimeStats.totalRequestsCountByTarget)) {
+                    metrics.gauge({
+                        name: `runtimeStats.requestsCount.target.${camelCase(target)}.total`,
+                        value: runtimeStats.totalRequestsCountByTarget[target],
+                    })
+                }
+            }, 1000)
+        }
+
+        // nosemgrep: javascript.express.security.audit.express-check-csurf-middleware-usage.express-check-csurf-middleware-usage
+        const app = express()
+
+        app.use(function runtimeStatsMiddleware (req, res, next) {
+            try {
+                const url = new URL(`${conf['SERVER_URL']}${req.url}`).pathname
+                const method = (req.method || 'get').toLowerCase()
+
+                const possibleXTargetValues = (conf[X_TARGET_OPTIONS_VAR_NAME] || '').split(',').filter(Boolean)
```

Comment:
Can we init this during middleware initialisation and not in runtime on each request?
We can also utilise Set for faster lookups

Reply:
And I think `xTargetOptions` or `targetOptions` is more aligned with other namings

Reply:
> Can we init this during middleware initialisation and not in runtime on each request? We can also utilise Set for faster lookups

Yes, done

Reply:
> And I think `xTargetOptions` or `targetOptions` is more aligned with other namings

Renamed